apicast/src/oauth/oidc.lua

```diff
@@ -0,0 +1,131 @@
+local jwt = require 'resty.jwt'
+local jwt_validators = require 'resty.jwt-validators'
+
+local lrucache = require 'resty.lrucache'
+local util = require 'util'
+
+local setmetatable = setmetatable
+local len = string.len
+local ngx_now = ngx.now
+local format = string.format
+
+local _M = {
+  cache_size = 10000,
+}
+
+function _M.reset()
+  _M.cache = lrucache.new(_M.cache_size)
+end
+
+_M.reset()
+
+local mt = {
+  __index = _M
+}
+
+function _M.new(service)
+  local issuer = service.oidc.issuer
+  local config = service.oidc.config or {}
+  local openid = config.openid or {}
+
+  return setmetatable({
+    service = service,
+    config = config,
+    issuer = issuer,
+    clock = ngx_now,
+    alg_whitelist = util.to_hash(openid.id_token_signing_alg_values_supported),
+    jwt_claims = {
+      nbf = jwt_validators.is_not_before(),
+      exp = jwt_validators.is_not_expired(),
+      iss = jwt_validators.equals_any_of({ issuer }),
+    },
+  }, mt)
+end
+
+local function timestamp_to_seconds_from_now(expiry, clock)
+  local time_now = (clock or ngx_now)()
+  local ttl = expiry and (expiry - time_now) or nil
+  return ttl
+end
+
+-- Formats the realm public key string into Public Key File (PKCS#8) format
+local function format_public_key(key)
+  if not key then
+    return nil, 'missing key'
+  end
+
+  local formatted_key = "-----BEGIN PUBLIC KEY-----\n"
+  local key_len = len(key)
+  for i=1,key_len,64 do
+    formatted_key = formatted_key..string.sub(key, i, i+63).."\n"
+  end
+  formatted_key = formatted_key.."-----END PUBLIC KEY-----"
+  return formatted_key
+end
+
+-- Parses the token - in this case we assume it's a JWT token
+-- Here we can extract authenticated user's claims or other information returned in the access_token
+-- or id_token by RH SSO
+local function parse_and_verify_token(self, jwt_token)
+  local cache = self.cache
+
+  if not cache then
+    return nil, 'not initialized'
+  end
+  local cache_key = format('%s:%s', self.service.id, jwt_token)
+
+  local jwt_obj = cache:get(cache_key)
+
+  if jwt_obj then
+    ngx.log(ngx.DEBUG, 'found JWT in cache for ', cache_key)
+    return jwt_obj
+  end
+
+  jwt_obj = jwt:load_jwt(jwt_token)
+
+  if not jwt_obj.valid then
+    ngx.log(ngx.WARN, jwt_obj.reason)
+    return jwt_obj, 'JWT not valid'
+  end
+
+  if not self.alg_whitelist[jwt_obj.header.alg] then
+    return jwt_obj, '[jwt] invalid alg'
+  end
+  -- TODO: this should be able to use DER format instead of PEM
+  local pubkey = format_public_key(self.config.public_key)
```

Comment:
Why? We support only Keycloak and I could not find this in the spec. So looks like this might be keycloak specific. 
If you have spec that defines this I'll be happy to validate the input.